courses/models.py

```diff
@@ -524,3 +526,164 @@ class ProgramEnrollmentAudit(AuditModel):
+        default=FinalGradeStatus.PENDING,
+        max_length=30,
+    )
+    course_run_paid_on_edx = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = ('user', 'course_run')
+
+    @classmethod
+    def get_audit_class(cls):
+        return FinalGradeAudit
+
+    def to_dict(self):
+        return serialize_model_object(self)
+
+    @property
+    def grade_percent(self):
+        """Returns the grade field value as a number out of 100 (or None if the value is None)"""
+        return self.grade * 100 if self.grade is not None else None
+
+    @classmethod
+    def get_frozen_users(cls, course_run):
+        """
+        Returns all the users with a final grade in a given course run
+        """
+        return list(
+            cls.objects.filter(
+                course_run=course_run, status=FinalGradeStatus.COMPLETE
+            ).values_list('user', flat=True)
+        )
+
+    def __str__(self):
+        return 'Grade in course "{course_id}", user "{user}", value {grade}'.format(
+            user=self.user.username,
+            grade=self.grade,
+            course_id=self.course_run.edx_course_key
+        )
+
+    def save(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        """Overridden method to run validation"""
+        self.full_clean()
+        return super().save(*args, **kwargs)
+
+
+class FinalGradeAudit(AuditModel):
+    """
+    Audit table for the Final Grade
+    """
+    final_grade = models.ForeignKey(FinalGrade, null=True, on_delete=models.SET_NULL)
+
+    @classmethod
+    def get_related_field_name(cls):
+        return 'final_grade'
+
+    def __str__(self):
+        return 'Grade audit for user "{user}", course "{course_id}"'.format(
+            user=self.final_grade.user,
+            course_id=self.final_grade.course_run.edx_course_key
+        )
+
+
+class CourseCertificate(TimestampedModel):
+    """
+    Model for storing course certificates
+    """
+    user = models.ForeignKey(User, models.SET_NULL, null=True, related_name='course_certificates')
+    course = models.ForeignKey(Course, models.SET_NULL, null=True)
+    hash = models.CharField(max_length=32, null=False, unique=True)
+
+    def save(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        """Overridden save method"""
+        if not self.hash:
+            self.hash = generate_md5(
+                '{}|{}'.format(self.user_id, self.course_id).encode('utf-8')
+            )
+        super().save(*args, **kwargs)
+
+    def __str__(self):
+        return 'Course Certificate for user={user}, course={course}, hash="{hash}"'.format(
+            user=self.user,
+            course=self.course.id,
+            hash=self.hash,
+        )
+
+
+class ProgramCertificate(TimestampedModel):
+    """
+    Model for storing program certificates
+    """
+    program = models.ForeignKey(Program, null=False, on_delete=models.CASCADE)
```

Comment:
How is this supposed to provide the duration of the program that this certificate is being generated for?

Reply:
did you get it above ?

Reply:
I am thinking to fetch the `ProgramEnrollment` for that particular `Program`  for the particular `user`.